Adding supervision tree visualizer

akoutmos opened the pull request:
This PR addresses half of #154. It supports rendering supervision trees top-down and left to right. You can render the supervisor by name or by PID. The links that are displayed are only between processes within the supervision tree and process links external to the supervision tree are omitted.

I haven't added tests yet, but will add them once I know I am on the right track :smile:. Thanks in advance for the review!

Some sample images for reference:

![image](https://user-images.githubusercontent.com/4753634/175823772-0f47fc50-13b6-4492-b96f-cd6d17437647.png)
![image](https://user-images.githubusercontent.com/4753634/175823869-c73c580c-0397-4af5-a7fb-7f0f9ef5ff4f.png)

josevalim commented on lib/kino/process.ex:
```diff
@@ -0,0 +1,251 @@
+    end
+  end
+
+  defmodule SupervisorGraphEdge do
```
Given these modules are private, we don't need a module, you can use "private" maps instead:

    defp graph_edge(node_1, node_2, relationship) do
      %{node_1: node_1, node_2: node_2, relationship: relationship}
    end

josevalim commented on lib/kino/process.ex:
```diff
@@ -0,0 +1,251 @@
+
+  """
+  @spec generate_supervision_tree(supervisor(), keyword()) :: Markdown.t()
+  def generate_supervision_tree(supervisor, opts \\ [])
```
Can we rename this to `sup_tree`? :) Or at least `supervision_tree` (but I would go with the former... it is fine short names here as it is an exploratory library.

josevalim commented on lib/kino/process.ex:
```diff
@@ -0,0 +1,251 @@
+  @spec generate_supervision_tree(supervisor(), keyword()) :: Markdown.t()
+  def generate_supervision_tree(supervisor, opts \\ [])
+
+  def generate_supervision_tree(named_supervisor, opts) when is_atom(named_supervisor) do
+    case Process.whereis(named_supervisor) do
```
Move the pid clause first and then do something like this:

```suggestion
  def generate_supervision_tree(named_supervisor, opts) do
    case GenServer.whereis(named_supervisor) do
```

This will also allow us to support via tuples. :)

josevalim commented on lib/kino/process.ex:
```diff
@@ -0,0 +1,251 @@
+    raise ArgumentError, "expected a PID, got: #{inspect(invalid_pid)}"
+  end
+
+  defp convert_direction(:td), do: "TD"
+  defp convert_direction(:lr), do: "LR"
```
```suggestion
  defp convert_direction(:top_down), do: "TD"
  defp convert_direction(:left_right), do: "LR"
```

?

josevalim commented on lib/kino/process.ex:
```diff
@@ -0,0 +1,251 @@
+  end
+
+  @doc """
+  Generates a Mermaid.js graph of a supervision tree. The provided supervisor can be either a named
```
First paragraph of the docs should be a short summary. :)

```suggestion
  Generates a Mermaid.js graph of a supervision tree.

  The provided supervisor can be either a named
```

jonatanklosko commented on lib/kino/process.ex:
```diff
@@ -0,0 +1,251 @@
+  ## Options
+
+    * `:direction` - defines the direction of the Mermaid.js graph. The
+      value can either be `:td` (top-down) or `:lr` (left to right) with
+      the default being `:td`. Optional
```
```suggestion
      value can either be `:td` (top-down) or `:lr` (left to right). Defaults
      to `:td`
```

akoutmos commented:
Thanks for the feedback! I think I addressed all of the comments regarding the new `sup_tree` function. I will work on implementing `app_tree` as well later today :smiley:.

jonatanklosko commented on lib/kino/process.ex:
````diff
@@ -0,0 +1,229 @@
+      ```
+      """)
+    catch
+      _, _ ->
+        raise ArgumentError, "the provided PID #{inspect(supervisor)} is not a supervisor"
````
Can we wrap the `Supervisor.which_children` call specifically? Otherwise it's not obvious where this comes from :)

akoutmos commented:
Almost done with the `app_tree` function as well. Should the `:application_master` processes get any kind of special formatting or is the image below good? I also left the thicker border on the application supervisor.

![image](https://user-images.githubusercontent.com/4753634/176070752-d258d3ef-67b1-47c2-9775-368303f11b53.png)

jonatanklosko commented on lib/kino/process.ex:
```diff
@@ -0,0 +1,305 @@
+                "the provided identifier #{inspect(supervisor)} does not reference a running process"
+      end
+
+    supervisor_children =
+      try do
+        Supervisor.which_children(supervisor)
+      catch
+        _, _ ->
+          raise ArgumentError, "the provided process #{inspect(supervisor)} is not a supervisor"
+      end
```
Sidenote: since this does GenServer call, it can easily crash or block the receiver, so even though we catch an error, there's likely an undesired side effect. @josevalim can we check for a supervisor upfront, perhaps by looking at `$initial_call`, or is that unreliable?

jonatanklosko commented on lib/kino/process.ex:
```diff
@@ -0,0 +1,305 @@
+    direction = direction_from_opts(opts)
+    edges = traverse_supervisor(root_supervisor)
+
+    [ancestor] =
+      root_supervisor
+      |> Process.info()
+      |> Keyword.get(:dictionary)
+      |> Keyword.get(:"$ancestors")
```
```suggestion
    {:dictionary, dictionary} = Process.info(root_supervisor, :dictionary)
    [ancestor] = dictionary[:"$ancestors"]
```

josevalim commented:
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

jonatanklosko commented:
@akoutmos fantastic job! :cat:

akoutmos commented:
Thanks for helping me get this feature in :)
